### PR TITLE
refactor: use Ordering inside of rbmap instead of lt.

### PR DESCRIPTION
**Pull request by DanielFabian**
I did go a bit crazy and made the `LT` to `Ordering` refactor. But it's big... The tests pass, so that's promising. 

But before we do the whole merge, it'd definitely make sense to read and double-check if my comparisons are as intended.

Maybe, a bit of hints like `inline` are in order, too. Any help is welcome.

### CODE REVIEW

File: src/Init/Data/Ord.lean [hunk 1]
```diff
@@ -37,3 +37,10 @@ instance : Ord Bool where
 
 instance : Ord String where
   compare x y := compareOfLessAndEq x y
+
+def USize.cmp (a b : USize) : Ordering := compare a.val.val b.val.val
```
Comment (Kha) [inline src/Init/Data/Ord.lean:41]: Missing `Ord (Fin n)`?

File: src/Init/Data/Ord.lean [hunk 2]
```diff
@@ -37,3 +37,10 @@ instance : Ord Bool where
...
+
+instance : Ord USize       := ⟨USize.cmp⟩
```
Comment (Kha) [inline src/Init/Data/Ord.lean:43]: Weird whitespace. Also, I think it's fine to define this inline with `where` as below.

File: src/Init/Prelude.lean [hunk 3]
```diff
@@ -828,7 +828,7 @@ instance : LT UInt32 where
 
 instance : LE UInt32 where
   le a b := LE.le a.val b.val
-
+
```
Comment (Kha) [inline src/Init/Prelude.lean:831]: trailing whitespace

File: src/Lean/Data/Name.lean [hunk 4]
```diff
@@ -65,26 +65,50 @@ def isSuffixOf : Name → Name → Bool
...
+  | str _ _ _,   num _ _ _   => Ordering.gt
+  | str p₁ n₁ _, str p₂ n₂ _ => 
+    match cmp p₁ p₂ with 
+    | Ordering.eq => compare n₁ n₂
+    | ord => ord
+    
+def lt (x y : Name) : Bool :=
+  match cmp x y with
+  | Ordering.lt => true
+  | _ => false
```
Comment (Kha) [inline src/Lean/Data/Name.lean:86]:
```suggestion
  cmp x y == Ordering.lt
```
should work too, right?

File: src/Lean/Data/Name.lean [hunk 5]
```diff
@@ -65,26 +65,50 @@ def isSuffixOf : Name → Name → Bool
...
-  | str n s _, str n' s' _ => s < s' || (s = s' && n.quickLtAux n')
-  | _,         _           => false
+def cmp : Name → Name → Ordering
+  | anonymous,   anonymous   => Ordering.eq
+  | anonymous,   _           => Ordering.lt
+  | _,           anonymous   => Ordering.gt
+  | num p₁ i₁ _, num p₂ i₂ _ => 
+    match cmp p₁ p₂ with  
+    | Ordering.eq => compare i₁ i₂
+    | ord => ord
```
Comment (Kha) [inline src/Lean/Data/Name.lean:75]: Should we have a function for this? In Haskell, this is simply the `Monoid` instance for `Ordering`, but we'd have to be a bit more careful with strictness...

File: src/Lean/Data/NameTrie.lean [hunk 6]
```diff
@@ -16,13 +16,19 @@ instance : ToString NamePart where
...
+  | _, _ => Ordering.gt
+
 def NamePart.lt : NamePart → NamePart → Bool
   | NamePart.str a, NamePart.str b => a < b
   | NamePart.num a, NamePart.num b => a < b
   | NamePart.num _, NamePart.str _ => true
   | _, _ => false
-
-def NameTrie (β : Type u) := PrefixTree NamePart β NamePart.lt
+
```
Comment (Kha) [inline src/Lean/Data/NameTrie.lean:30]: trailing whitespace

File: src/Std/Data/RBMap.lean [hunk 7]
```diff
@@ -94,31 +94,32 @@ def isBlack : RBNode α β → Bool
...
 @[specialize] def ins : RBNode α β → (k : α) → β k → RBNode α β
   | leaf,               kx, vx => node red leaf kx vx leaf
   | node red a ky vy b, kx, vx =>
-     if lt kx ky then node red (ins a kx vx) ky vy b
-     else if lt ky kx then node red a ky vy (ins b kx vx)
-     else node red a kx vx b
+    match cmp kx ky with
+    | Ordering.lt => node red (ins a kx vx) ky vy b
+    | Ordering.gt => node red a ky vy (ins b kx vx)
+    | Ordering.eq => node red a kx vx b
```
Comment (Kha) [inline src/Std/Data/RBMap.lean:105]: So much better :grin:

**Review by Kha — APPROVED** [review]
Looks good to me mostly

File: src/Lean/Data/Name.lean (hunk 4 above)
Comment (Kha) [reply src/Lean/Data/Name.lean:86]: I guess it doesn't, but then we should make it work

File: src/Init/Data/Ord.lean (hunk 2 above)
Comment (DanielFabian) [reply src/Init/Data/Ord.lean:43]: done

File: src/Init/Prelude.lean (hunk 3 above)
Comment (DanielFabian) [reply src/Init/Prelude.lean:831]: done

File: src/Lean/Data/Name.lean (hunk 4 above)
Comment (DanielFabian) [reply src/Lean/Data/Name.lean:86]: sure it works, just needed the `BEq` deriving. I call it a case of "Wald vor lauter Bäumen nicht gesehen".

File: src/Lean/Data/NameTrie.lean (hunk 6 above)
Comment (DanielFabian) [reply src/Lean/Data/NameTrie.lean:30]: done

File: src/Std/Data/RBMap.lean (hunk 7 above)
Comment (DanielFabian) [reply src/Std/Data/RBMap.lean:105]: yep

File: src/Init/Data/Ord.lean (hunk 1 above)
Comment (DanielFabian) [reply src/Init/Data/Ord.lean:41]: done

File: src/Lean/Data/Name.lean (hunk 5 above)
Comment (DanielFabian) [reply src/Lean/Data/Name.lean:75]: so I guess we could have a `combine : Ordering -> Thunk Ordering -> Ordering`, but is it really so much easier?

Comment (DanielFabian) [reply src/Lean/Data/Name.lean:75]: sometimes, I feel like the 3-line solution is ok.

Comment (Kha) [reply src/Lean/Data/Name.lean:75]: Yeah, it's not too bad :)

Comment (pnwamk) [reply src/Lean/Data/Name.lean:75]: I would imagine this pattern is common enough I would probably find myself reaching for something like a default `Ord` instance for products to handle it (assuming the comparisons are inlined at least... perhaps a variadic macro could work too?)
```lean
instance [Ord α] [Ord β] : Ord (α × β) where
  compare | (a₁, b₁), (a₂, b₂) =>
            match compare a₁ a₂ with
            | Ordering.eq => compare b₁ b₂
            | ord => ord

#eval compare (1,2,3) (1,2,4)
```

Comment (pnwamk) [reply src/Lean/Data/Name.lean:75]: Ah but the first use in that example uses `cmp` and the second usage is to `compare`... so perhaps this is less applicable here.

**Comment by Kha** [comment]
Unless I did something wrong on my side, this PR does in fact introduce a large compilation time regression: http://speedcenter.informatik.kit.edu/velcom/compare/bab50d33-13a1-40c4-85a0-fda5f93f8b22/to/a1ef5809-7e78-4e6e-b3e7-9518f654039b?hash1=14c5ec559bcfd9fc7b1d93ff6cf04ef2e70d9209
I don't have time right now to look into it further.

**Comment by DanielFabian** [comment]
the two really big changes seem to be those:
image
From how I read it, it's generating a ton more code. The pref doesn't actually seem much affected if you look at the timing tests. My guess would actually be more like like we're generating tons of new definitions. I wonder if it's got to do with the `below` construction? 

Although it does feel like a lot even then. It's not like we've got that many inductive predicates, is it?

**Comment by Kha** [comment]
No, the comparison commit also includes the `below` commit. It has to be changes from this commit.

File: src/Lean/Data/Name.lean (hunk 5 above)
Comment (Kha) [reply src/Lean/Data/Name.lean:75]: Yes, outside of recursive `Ord` instances, just using pairs seems like a nice solution

**Comment by Kha** [comment]
Since even importing got significantly slower, I'm assuming that the issue is not that the stdlib got more costly to process, but that Lean itself got slower. The prime suspect would be (bad compilation of) a central function like `Name.quickCmp`, though I don't immediately see anything wrong with it.

**Comment by DanielFabian** [comment]
@Kha dumb question? how do 3-way inductives get translated to C vs 2-way ones? I'd expect a 2-way case becomes effectively a `compare non-zero` or something like that?

**Comment by DanielFabian** [comment]
also, can we see the generated code for `quickCmp`?

**Comment by pnwamk** [comment]
@DanielFabian Do you mean the generated C source for it? Here's that if so: https://gist.github.com/pnwamk/b2dd37d06f559c868383d2ade81a9452#file-lean_data_name-c-L1780

**Comment by DanielFabian** [comment]
that's a lot of code for a string and number comparison, lol

File: src/Lean/Data/Name.lean (hunk 5 above)
Comment (leodemoura) [reply src/Lean/Data/Name.lean:75]: The `import` time improved a bit after I added `@[inline]` at
```lean
@[inline] def compareOfLessAndEq {α} ...
```
but it is still slower.

**Comment by leodemoura** [comment]
The import time improved a bit after I added @[inline] at
```lean
@[inline] def compareOfLessAndEq {α} ...
```
but it is still slower. After this change, the code generated for `l_Lean_Name_quickCmpAux` is the one I expected.

**Comment by leodemoura** [comment]
The second perf problem was due to the following instance
```lean
 instance : Ord USize where
-  compare x y := compare x.val y.val
+  compare x y := compareOfLessAndEq x y
```

**Comment by leodemoura** [comment]
`import` is now as fast as before after these changes. I will merge this PR.

**Comment by leodemoura** [comment]
merged

**Pull request closed without merging**